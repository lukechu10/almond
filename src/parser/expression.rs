//! Parsing for JS expressions.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom_locate::position;

/// Alias for `parse_expr_bp(s, 0)`.
pub fn parse_expr(s: Span) -> ParseResult<Node> {
    context("expression", |s| parse_expr_bp(s, 0, false))(s)
}

/// Alias for `parse_expr_bp(s, 1)`. Should be used when parsing expressions in expression lists.
/// This prevents matching the sequence (`,`) operator.
pub fn parse_expr_no_seq(s: Span) -> ParseResult<Node> {
    context("expression no seq", |s| parse_expr_bp(s, 1, false))(s)
}

/// Parse an atomic expression — either a single token that is an
// expression, an expression started by a keyword like `function`.
pub fn parse_primary_expr(s: Span) -> ParseResult<Node> {
    alt((
        parse_identifier,
        literal::parse_literal,
        parse_this_expr,
        parse_function_expr,
        parse_paren_expr,
    ))(s)
}

/// Parse an atomic expression — either a single token that is an
/// expression, an expression started by a keyword like `function`.
/// This variant of `parse_primary_expr` allows parsing an identifier that is a reserved name.
/// This method should be used instead of `parse_primary_expr` when preceding token is `.` operator (for member expression).
/// # Spec
/// http://www.ecma-international.org/ecma-262/#sec-property-accessors
pub fn parse_primary_expr_allow_reserved(s: Span) -> ParseResult<Node> {
    alt((
        parse_identifier_name, // note that this is different from `parse_identifier` which does not allow reserved name.
        literal::parse_literal,
        parse_this_expr,
        parse_function_expr,
        parse_paren_expr,
    ))(s)
}

pub fn parse_this_expr(s: Span) -> ParseResult<Node> {
    map(
        spanned(ws0(pair(tag("this"), not(identifier_continue)))),
        |(_, start, end)| NodeKind::ThisExpression.with_pos(start, end),
    )(s)
}

pub fn parse_paren_expr(s: Span) -> ParseResult<Node> {
    context(
        "paren expression",
        delimited(ws0(char('(')), parse_expr, ws0(char(')'))),
    )(s)
}

fn parse_opt_expr_in_list(s: Span) -> ParseResult<Option<Node>> {
    alt((value(None, peek(char(','))), map(parse_expr_no_seq, Some)))(s)
}

pub fn parse_expr_list_with_opt_expr(s: Span) -> ParseResult<Vec<Option<Node>>> {
    context(
        "expression list with optional expression",
        terminated(
            separated_list0(ws0(char(',')), parse_opt_expr_in_list),
            // trailing comma
            ws0(opt(char(','))),
        ),
    )(s)
}

pub fn parse_expr_list(s: Span) -> ParseResult<Vec<Node>> {
    context(
        "expression list",
        terminated(
            separated_list0(ws0(char(',')), parse_expr_no_seq),
            // trailing comma
            ws0(opt(char(','))),
        ),
    )(s)
}

/// Pratt parsing for prefix operators. Called in `parse_expr_bp`.
fn parse_prefix_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, (prefix_op, BindingPower(_, right_bp))) = parse_prefix_operator(s)?;
    let (s, rhs) = parse_expr_bp(
        s, right_bp, /* reserved is only used after `.` operator */ false,
    )?;

    let (mut s, mut end) = position(s)?;

    let node_kind = match prefix_op {
        PrefixOperator::Unary(prefix_op) => NodeKind::UnaryExpression {
            argument: Box::new(rhs),
            operator: prefix_op,
            prefix: true,
        },
        PrefixOperator::Update(prefix_op) => NodeKind::UpdateExpression {
            argument: Box::new(rhs),
            operator: prefix_op,
            prefix: true,
        },
        PrefixOperator::New => {
            let (s_tmp, arguments) = opt(delimited(ws0(char('(')), parse_expr_list, char(')')))(s)?;
            s = s_tmp;
            let (s_tmp, end_tmp) = position(s)?;
            s = s_tmp;
            end = end_tmp;

            let (s_tmp, _) = sp0(s)?;
            s = s_tmp;

            NodeKind::NewExpression {
                callee: Box::new(rhs),
                arguments: arguments.unwrap_or_default(),
            }
        }
        PrefixOperator::Await => NodeKind::AwaitExpression {
            argument: Box::new(rhs),
        },
    };
    Ok((s, node_kind.with_pos(start, end)))
}

/// Pratt parsing for expressions with operator precedence.
/// Check out [https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html](https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html) to see how Pratt parsing works.
/// # Params
/// * `min_bp` - The minimal binding power to accept.
/// * `allow_reserved` - If `true`, will call `parse_primary_expr_allow_reserved` instead of `parse_primary_expr`. In almost all cases, this should be `false`.
pub fn parse_expr_bp(s: Span, min_bp: i32, allow_reserved: bool) -> ParseResult<Node> {
    let (mut s, mut lhs) = alt((
        parse_prefix_expr,
        if allow_reserved {
            parse_primary_expr_allow_reserved
        } else {
            parse_primary_expr
        },
    ))(s)?;

    loop {
        if let Ok((s_tmp, (postfix_op, BindingPower(left_bp, _), mut end))) =
            parse_postfix_operator(s)
        {
            if left_bp < min_bp {
                break;
            }
            s = s_tmp;
            let start = lhs.start.clone();

            let node_kind = match postfix_op {
                PostfixOperator::Update(postfix_op) => NodeKind::UpdateExpression {
                    argument: Box::new(lhs),
                    operator: postfix_op,
                    prefix: false,
                },
                PostfixOperator::ComputedMember => {
                    // array access
                    let (s_tmp, property) = terminated(parse_expr, char(']'))(s)?;
                    s = s_tmp;

                    let (s_tmp, end_tmp) = position(s)?;
                    s = s_tmp;
                    end = end_tmp;

                    let (s_tmp, _) = sp0(s)?;
                    s = s_tmp;

                    NodeKind::MemberExpression {
                        object: Box::new(lhs),
                        property: Box::new(property),
                        computed: true,
                    }
                }
                PostfixOperator::FuncCall => {
                    // array access
                    let (s_tmp, arguments) = terminated(parse_expr_list, char(')'))(s)?;
                    s = s_tmp;

                    let (s_tmp, end_tmp) = position(s)?;
                    s = s_tmp;
                    end = end_tmp;

                    let (s_tmp, _) = sp0(s)?;
                    s = s_tmp;

                    NodeKind::CallExpression {
                        callee: Box::new(lhs),
                        arguments,
                    }
                }
            };

            lhs = node_kind.with_pos(start, end);

            continue;
        }

        // do not override s just yet
        let (s_tmp, (op, BindingPower(left_bp, right_bp))) = match parse_infix_operator(s) {
            Ok(res) => res,
            Err(_) => break, // do not return from function, just break from loop.
        };

        if left_bp < min_bp {
            break;
        }

        // ok, now we can override s
        s = s_tmp;

        if let InfixOperator::TernaryOperator = op {
            let (s_tmp, mhs) = parse_expr_bp(s, right_bp, false)?;
            s = s_tmp;

            let (s_tmp, _) = ws0(tag(":"))(s)?;
            s = s_tmp;

            let (s_tmp, rhs) = parse_expr_bp(s, right_bp, false)?;
            s = s_tmp;

            let start = lhs.start.clone();
            let end = rhs.end.clone();

            let node_kind = NodeKind::ConditionalExpression {
                test: Box::new(lhs),
                consequent: Box::new(mhs),
                alternate: Box::new(rhs),
            };

            lhs = node_kind.with_pos(start, end);
            continue;
        }

        let (s_tmp, rhs) = parse_expr_bp(s, right_bp, op == InfixOperator::DotOperator)?;
        s = s_tmp;

        let start = lhs.start.clone();
        let end = rhs.end.clone();

        let node_kind = match op {
            InfixOperator::Binary(op) => NodeKind::BinaryExpression {
                left: Box::new(lhs),
                right: Box::new(rhs),
                operator: op,
            },
            InfixOperator::Logical(op) => NodeKind::LogicalExpression {
                left: Box::new(lhs),
                right: Box::new(rhs),
                operator: op,
            },
            InfixOperator::Assignment(op) => NodeKind::AssignmentExpression {
                left: Box::new(lhs),
                right: Box::new(rhs),
                operator: op,
            },
            InfixOperator::DotOperator => NodeKind::MemberExpression {
                object: Box::new(lhs),
                property: Box::new(rhs),
                computed: false,
            },
            InfixOperator::SequenceOperator => NodeKind::SequenceExpression {
                expressions: match &lhs.kind {
                    NodeKind::SequenceExpression { expressions } => {
                        let mut expressions = expressions.clone();
                        expressions.push(rhs);
                        expressions
                    }
                    _ => vec![lhs, rhs],
                },
            },
            InfixOperator::TernaryOperator => unreachable!("handled earlier"),
        };
        lhs = node_kind.with_pos(start, end);
    }

    Ok((s, lhs))
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn smoke_test_this_expr() {
        parse_this_expr("this".into()).unwrap();
        parse_this_expr("notthis".into()).unwrap_err();
    }

    #[test]
    fn smoke_test_primary_expr() {
        parse_primary_expr("this".into()).unwrap();
        parse_primary_expr("myVar".into()).unwrap();
    }

    #[test]
    fn test_identifier_expr() {
        assert_json_snapshot!(parse_expr("myIdentifier".into()).unwrap().1);
    }

    #[test]
    fn test_expr_bp_guard_in() {
        assert_json_snapshot!(
            parse_expr_bp("myIdentifier in foo".into(), 25, false)
                .unwrap()
                .1
        );
        // should only parse myIdentifier
    }

    #[test]
    fn test_paren_expr() {
        assert_json_snapshot!(parse_expr("(1)".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("(((1)))".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("(((1 + 1)))".into()).unwrap().1);
    }

    #[test]
    fn test_member_expr() {
        assert_json_snapshot!(parse_expr("a.b".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("a.b.c".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("a[1]".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("a[0]".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("a[[]]".into()).unwrap().1);
    }

    #[test]
    fn test_new_expr() {
        assert_json_snapshot!(parse_expr("new Array()".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("new Array".into()).unwrap().1); // paren optional
        assert_json_snapshot!(parse_expr("new Array(1)".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("new Array(1,)".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("new Array(1,2)".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("new Foo.Bar(true)".into()).unwrap().1);
    }

    #[test]
    fn test_call_expr() {
        assert_json_snapshot!(parse_expr("foo()".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("foo.bar()".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("foo.bar.baz()".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("foo.bar(baz)".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("foo.bar(baz, 1, 2, 3)".into()).unwrap().1);

        assert_json_snapshot!(
            parse_expr("console.log(\"Hello World!\")".into())
                .unwrap()
                .1
        );
    }

    #[test]
    fn test_sequence_expr() {
        assert_json_snapshot!(parse_expr("1,2,3".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("(1,2,3)".into()).unwrap().1);
    }

    #[test]
    fn test_ternary_expr() {
        assert_json_snapshot!(parse_expr("true ? x : y".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("a ? x : b ? y : z".into()).unwrap().1);
        // should be parsed as a ? x : (b ? y : z)
    }

    #[test]
    fn test_expr_bp_infix() {
        assert_json_snapshot!(parse_expr("1 + 2".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 - 2 - 3".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 * 2 + 3".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 + 2 * 3".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 * 2 + 3 * 4".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("(1 + 2) * 3".into()).unwrap().1);

        assert_json_snapshot!(parse_expr("true && false".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x < y".into()).unwrap().1);

        // equality
        assert_json_snapshot!(parse_expr("x == 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x != 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x === 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x !== 1".into()).unwrap().1);

        // relational
        assert_json_snapshot!(parse_expr("x < 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x > 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x <= 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x >= 1".into()).unwrap().1);

        // bitwise
        assert_json_snapshot!(parse_expr("x << 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x >> 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x >>> 1".into()).unwrap().1);

        // logical
        assert_json_snapshot!(parse_expr("x && y".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x || y".into()).unwrap().1);

        // exponentiation (right associative)
        assert_json_snapshot!("exponentiation", parse_expr("x ** y".into()).unwrap().1);
        assert_json_snapshot!(
            "exponentiation-2",
            parse_expr("x ** y ** z".into()).unwrap().1
        );
    }

    #[test]
    fn test_expr_bp_infix_str_concat() {
        assert_json_snapshot!(parse_expr(r#""a" + "b""#.into()).unwrap().1);
        assert_json_snapshot!(parse_expr(r#""a" + 123"#.into()).unwrap().1);
        assert_json_snapshot!(parse_expr(r#""a" + {}"#.into()).unwrap().1);
        assert_json_snapshot!(parse_expr(r#"[] + {}"#.into()).unwrap().1);
    }

    #[test]
    fn test_expr_bp_prefix() {
        assert_json_snapshot!(parse_expr("-1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("+1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("+(+1)".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 + -2".into()).unwrap().1); // same as 1 + (-2)
        assert_json_snapshot!(parse_expr("++x".into()).unwrap().1);

        assert_json_snapshot!(parse_expr("typeof x".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("typeof module === \"object\"".into()).unwrap().1);

        assert_json_snapshot!(parse_expr("!x".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("!(x && y)".into()).unwrap().1);

        assert_json_snapshot!(parse_expr("await foo".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("await foo".into()).unwrap().1);
    }

    #[test]
    fn test_expr_bp_postfix() {
        assert_json_snapshot!(parse_expr("x++".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x++\t".into()).unwrap().1); // make sure end pos does not include trailing whitespace
        assert_json_snapshot!(parse_expr("x--".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x++ + 1".into()).unwrap().1); // ++ binds tighter than +
    }

    #[test]
    fn test_expr_bp_assignment() {
        assert_json_snapshot!(parse_expr("x = 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x = 1 + 2".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x += 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x += y += 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x += x * x".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x = a ? b : c".into()).unwrap().1);
        assert_json_snapshot!(
            parse_expr("x = a ? myFunc : function () { return c; }".into())
                .unwrap()
                .1
        );

        assert_json_snapshot!(parse_expr("hello.value = \"world\"".into()).unwrap().1);
        assert_json_snapshot!(
            parse_expr("myFunc = function () { return 0; }".into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(parse_expr("a = b = c;".into()).unwrap().1);

        assert_json_snapshot!(parse_expr("x -= x * x".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x **= y".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x *= x * x".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x /= x * x".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x %= x * x".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x <<= 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x >>= 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x >>>= 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x |= 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x ^= 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x &= 1".into()).unwrap().1);

        assert_json_snapshot!(parse_expr("x[1] = a".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x[\"foo\"] = a".into()).unwrap().1);

        assert_json_snapshot!(
            parse_expr(
                r#"identifier = "(?:\\\\[\\da-fA-F]{1,6}" + whitespace +
				"?|\\\\[^\\r\\n\\f]|[\\w-]|[^\0-\\x7f])+""#
                    .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_expr_bp_member_expr() {
        assert_json_snapshot!(parse_expr("x.y".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x.y.z".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x.y[z]".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x.y[\"z\"]".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x.y[arr.length - 1]".into()).unwrap().1);

        assert_json_snapshot!(
            parse_expr("x.y()\n// abc\n/* 123 */\n.z()".into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_expr("x.y(123)\n// abc\n/* 123 */\n.z(abc)".into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_expr(
                r#"test
                    .then(fn)
                    // properties can be reserved words ("catch" keyword)
                    .catch(function (error) {
                        console.error(error);
                    });"#
                    .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_expr_bp_no_seq() {
        assert_json_snapshot!(parse_expr_no_seq("x, y".into()).unwrap().1); // should not parse ", y"
        assert_json_snapshot!(
            parse_expr_no_seq(
                "test() ? function () { return 0; } : function() { return 1; }, y".into()
            )
            .unwrap()
            .1
        ); // should not parse ", y"
    }
}
