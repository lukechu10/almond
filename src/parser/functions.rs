//! Parsing for JS functions and programs.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::InputLength;
use nom_locate::position;

/// Parses a complete JS program
pub fn parse_program(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, _) = sp0(s)?; // eat all preceding whitespace

    let (s, body) = parse_function_body_inner(s)?;
    if s.input_len() != 0 {
        panic!("The source code was not completely parsed. This is a bug with the parser.\nUnparsed code starting on line {}", s.location_line());
    }

    let (s, end) = position(s)?; // Program loc should include all trailing whitespace
    Ok((s, NodeKind::Program { body }.with_pos(start, end)))
}

pub fn parse_declaration(s: Span) -> ParseResult<Node> {
    parse_function_declaration(s)
}

pub fn parse_function_declaration(s: Span) -> ParseResult<Node> {
    let parse_function_declaration_signature = pair(
        separated_pair(
            ws0(opt(keyword_async)),
            ws1(keyword_function),
            parse_identifier,
        ),
        delimited(ws0(tag("(")), parse_formal_param_list, ws0(tag(")"))),
    );

    context(
        "function declaration",
        map(
            spanned(pair(
                parse_function_declaration_signature,
                parse_function_body,
            )),
            |((((is_async, id), params), body), start, end)| {
                NodeKind::FunctionDeclaration {
                    function: Function {
                        id: Box::new(Some(id)),
                        params,
                        body: Box::new(body),
                        is_async: is_async.is_some(),
                    },
                }
                .with_pos(start, end)
            },
        ),
    )(s)
}

pub fn parse_function_expr(s: Span) -> ParseResult<Node> {
    let parse_function_expr_signature = pair(
        separated_pair(
            ws0(opt(keyword_async)),
            ws0(keyword_function),
            opt(parse_identifier),
        ),
        delimited(ws0(tag("(")), parse_formal_param_list, ws0(tag(")"))),
    );

    context(
        "function expression",
        map(
            spanned(pair(parse_function_expr_signature, parse_function_body)),
            |((((is_async, id), params), body), start, end)| {
                NodeKind::FunctionExpression {
                    function: Function {
                        id: Box::new(id),
                        params,
                        body: Box::new(body),
                        is_async: is_async.is_some(),
                    },
                }
                .with_pos(start, end)
            },
        ),
    )(s)
}

pub fn parse_function_body(s: Span) -> ParseResult<Node> {
    context(
        "function body",
        map(
            spanned(delimited(
                ws0(tag("{")),
                parse_function_body_inner,
                ws0(tag("}")),
            )),
            |(body, start, end)| NodeKind::BlockStatement { body }.with_pos(start, end),
        ),
    )(s)
}

pub fn parse_function_body_inner(s: Span) -> ParseResult<Vec<Node>> {
    let parse_directive_list = many0(parse_directive);
    let parse_source_element_list = many0(parse_source_elem);

    context(
        "function body inner",
        map(
            pair(parse_directive_list, parse_source_element_list),
            |(mut directives, stmts)| {
                directives.extend(stmts);
                directives
            },
        ),
    )(s)
}

pub fn parse_formal_param_list(s: Span) -> ParseResult<Vec<Node>> {
    terminated(
        separated_list0(ws0(char(',')), parse_formal_param),
        // trailing comma
        opt(ws0(char(','))),
    )(s)
}

pub fn parse_formal_param(s: Span) -> ParseResult<Node> {
    parse_identifier(s)
}

/// Example: `"use strict";`
pub fn parse_directive(s: Span) -> ParseResult<Node> {
    let parse_directive_literal = spanned(terminated(ws0(string_lit), opt(ws0(semi))));
    map(parse_directive_literal, |(node, start, end)| {
        match &node.kind {
            NodeKind::Literal { value } => match value {
                LiteralValue::String(directive) => {
                    let directive = directive.clone();
                    NodeKind::ExpressionStatement {
                        expression: Box::new(node),
                        directive: Some(directive),
                    }
                    .with_pos(start, end)
                }
                _ => unreachable!(
                    "string_lit should return a literal with value LiteralValue::String"
                ),
            },
            _ => unreachable!("string_lit should return a Node::Identifier"),
        }
    })(s)
}

pub fn parse_source_elem(s: Span) -> ParseResult<Node> {
    alt((parse_declaration, parse_stmt))(s)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn test_directive() {
        assert_json_snapshot!(parse_directive("\"use strict\"".into()).unwrap().1);
    }

    #[test]
    fn test_function_declaration() {
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    "use strict"
                    return 1
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a() {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a(param1, param2) {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a(param1, param2,) {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    "use strict"
                    "also directive"
                    a;
                    "not directive"
                    return 1
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function fib() {
                    if (x === 0) return 0;
                    else if (x === 1) return 1;
                    else return fib(x - 1) + fib(x - 2);
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"function foo() {
                    // stuff
                }; foo();"# // note the `;` after the function declaration. Should be parsed as an `EmptyStatement`.
                    .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_nested_function_declaration() {
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    function b() {
                        a();
                    }
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_function_expr() {
        parse_function_declaration(
            r#"function a() {
                // function expressions should be wrapped up in parenthesis
                function () {
                    a();
                }
            }"#
            .into(),
        )
        .unwrap_err();

        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    // anonymous function:
                    (function () {
                        a();
                    })
                }"#
                .into()
            )
            .unwrap()
            .1
        );

        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    // function expression
                    0, function b() {}
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn parse_immediately_invoked_function_expr() {
        assert_json_snapshot!(
            parse_stmt(
                r#"(function () {
                    123; // something
                })()"#
                    .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_program() {
        assert_json_snapshot!(
            parse_program(
                r#"var x = 1;
                function foo() {
                    return x;
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"var x = 1;
                function foo() {
                    if (x) {
                        return x;
                    }
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"
                function fib() {
                    if (x === 0) return 0;
                    else if (x === 1) return 1;
                    else return fib(x - 1) + fib(x - 2);
                }
                var x = fib(10);"#
                    .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"function foo() {
                    test;
                    // abc "123"-
                    // test
                    foo();
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_empty_function() {
        assert_json_snapshot!(
            parse_program(
                r#"function foo(callback) {
                    // A👍a
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn parse_program_immediately_invoked_function_expr() {
        assert_json_snapshot!(
            parse_program(
                r#"(function () {
                    123; // something
                })()"#
                    .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"(function (global, factory) {
                    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
                        typeof define === 'function' && define.amd ? define(['exports'], factory) :
                            (global = global || self, factory(global.Library = {}));
                }(this, (function (exports) {})));"#
                    .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_async_function() {
        assert_json_snapshot!(
            parse_program(
                r#"async function foo(callback) {
                    await test();
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }
}
