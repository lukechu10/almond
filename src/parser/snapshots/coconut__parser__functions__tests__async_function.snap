---
source: src/parser/functions.rs
expression: "parse_program(r#\"async function foo(callback) {\n                    await test();\n                }\"#.into()).unwrap().1"
---
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "foo",
        "start": 15,
        "end": 18
      },
      "params": [
        {
          "type": "Identifier",
          "name": "callback",
          "start": 19,
          "end": 27
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "AwaitExpression",
              "argument": {
                "type": "CallExpression",
                "callee": {
                  "type": "Identifier",
                  "name": "test",
                  "start": 57,
                  "end": 61
                },
                "arguments": [],
                "start": 57,
                "end": 63
              },
              "start": 51,
              "end": 63
            },
            "start": 51,
            "end": 64
          }
        ],
        "start": 29,
        "end": 82
      },
      "async": true,
      "start": 0,
      "end": 82
    }
  ],
  "start": 0,
  "end": 82
}
